plugins/snapshots/blockfile/blockfile.go

```diff
@@ -473,5 +488,25 @@ func copyFileWithSync(target, source string) error {
 	defer tgt.Sync()
 
 	_, err = io.Copy(tgt, src)
+	if err != nil {
+		return fmt.Errorf("failed to copy file to  target %s: %w", target, err)
+	}
+	if copySparse {
+		if err = fallocate(target); err != nil {
+			return fmt.Errorf("fallocate file %s to sparse file failed: %w", target, err)
+		}
+	}
+
 	return err
 }
+
+// fallocate  can make existing files sparse on supported file systems.
```

Comment:
‘Sparse’ means hole which doesn’t take filesystem space and is supposed to be generated by using truncate() or punch hole or truncate command. ‘Fallocate’ generates unwritten extents, which allocates space but leave those extents as unwritten status.

Reply:
I had a brief talk with @hsiangkao on slack, and I've re-implemented the sparseCopy logic from the original patch.